[WGSL] Add the type implementations to the type checker

tadeuzagallo opened the pull request:
#### 467cdc607241bf1ea213a6d222761095c056f1f1
<pre>
[WGSL] Add the type implementations to the type checker
<a href="https://bugs.webkit.org/show_bug.cgi?id=251611">https://bugs.webkit.org/show_bug.cgi?id=251611</a>
&lt;rdar://problem/104966774&gt;

Reviewed by Myles C. Maxfield.

The original commit only had a placeholder empty `Type` struct. This actually
adds the type definitions that we&apos;ll need to start implement the type checking.

* Source/WebGPU/WGSL/AST/ASTStringDumper.cpp:
(WGSL::AST::dumpNode): Deleted.
* Source/WebGPU/WGSL/AST/ASTStringDumper.h:
(WGSL::AST::dumpNode):
* Source/WebGPU/WGSL/AST/ASTTypeName.h:
* Source/WebGPU/WGSL/TypeCheck.cpp:
(WGSL::TypeChecker::TypeChecker):
(WGSL::TypeChecker::visit):
(WGSL::TypeChecker::infer):
(WGSL::TypeChecker::allocateType):
* Source/WebGPU/WGSL/Types.cpp: Added.
(WGSL::printInternal):
* Source/WebGPU/WGSL/Types.h: Added.
* Source/WebGPU/WebGPU.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/259805@main">https://commits.webkit.org/259805@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ff7705230b6129e897b3e2a61f387eac2de59df5

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/105913 "4 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/14981 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/38759 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/115107 "Built successfully") | 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/16399 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/6172 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/98141 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/114860 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/111663 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/12481 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/95452 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/40030 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/94346 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/27096 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/81694 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/8244 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/28450 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/8727 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/5035 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/14357 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/47990 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6787 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/10280 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

litherum commented on Source/WebGPU/WGSL/TypeCheck.cpp:
```diff
@@ -124,10 +126,55 @@ class UnificationContext {
+    : m_shaderModule(shaderModule)
+    , m_typeConstrutors(AST::ParameterizedTypeName::NumberOfBaseTypes)
+{
+    m_abstractInt = allocateType<Primitive>(Primitive::AbstractInt);
+    m_abstractFloat = allocateType<Primitive>(Primitive::AbstractFloat);
+    m_void = allocateType<Primitive>(Primitive::Void);
+    m_bool = allocateType<Primitive>(Primitive::Bool);
+    m_i32 = allocateType<Primitive>(Primitive::I32);
+    m_u32 = allocateType<Primitive>(Primitive::U32);
+    m_f32 = allocateType<Primitive>(Primitive::F32);
```
In WHLSL, we had these defined in a separate class (named `Intrinsics`) whose whole job was to implement the built-in types / functions. Should we do that too?

tadeuzagallo replied:
That sounds good, I considered something similar but ended up not doing it. I'll land this and post another PR to refactor it.

litherum commented on Source/WebGPU/WGSL/TypeCheck.cpp:
```diff
@@ -124,10 +126,55 @@ class UnificationContext {
+    f(Mat4x3, Matrix, static_cast<uint8_t>(4), static_cast<uint8_t>(3)) \
+    f(Mat4x4, Matrix, static_cast<uint8_t>(4), static_cast<uint8_t>(4))
+
+#define DEFINE_TYPE_CONSTRUCTOR(base, type, ...) \
```
Can't this be done with templates instead?

tadeuzagallo replied:
Good point, seems like it should work! I'll refactor it when I move to Intrinsics.

litherum commented on Source/WebGPU/WGSL/Types.cpp:
```diff
@@ -0,0 +1,68 @@
+
+void printInternal(PrintStream& out, const Type& type)
+{
+    WTF::switchOn(type,
```
lol WGSL is such a simple language

litherum commented on Source/WebGPU/WGSL/Types.cpp:
```diff
@@ -0,0 +1,68 @@
+        },
+        [&](const Primitive& primitive) {
+            switch (primitive.kind) {
+#define PRIMITIVE_CASE(kind, name) \
```
Is this really better than just listing them all out?

If we update the list in the future and forget to update this place, we'll get a compilation error...

tadeuzagallo replied:
Are you referring to the macro or the `switchOn`?

For the macro, I personally prefer it over having an enum and then having the switch here, but I don't feel strongly about it, just think it's easier to keep the definition + string representation in the same place.

The reason I asked about the `switchOn` is because I don't see how updating the macro would break this.

litherum commented on Source/WebGPU/WGSL/Types.h:
```diff
@@ -0,0 +1,105 @@
+struct Array;
+struct Struct;
+struct Function;
+
```
Pretty wild that we can actually just list all the types. 🆒

litherum commented on Source/WebGPU/WGSL/Types.h:
```diff
@@ -0,0 +1,105 @@
+};
+
+struct Vector {
+    Type* element;
```
Is it really okay that these are non-owning? What's guaranteeing we don't have UAFs here?

tadeuzagallo replied:
The TypeChecker keeps all the types alive. Once we start storing them in the AST we'll need to move them after type checking into the (yet to be created) class that will hold both the AST and the types.

tadeuzagallo committed:
[WGSL] Add the type implementations to the type checker
https://bugs.webkit.org/show_bug.cgi?id=251611
<rdar://problem/104966774>

Reviewed by Myles C. Maxfield.

The original commit only had a placeholder empty `Type` struct. This actually
adds the type definitions that we'll need to start implement the type checking.

* Source/WebGPU/WGSL/AST/ASTStringDumper.cpp:
(WGSL::AST::dumpNode): Deleted.
* Source/WebGPU/WGSL/AST/ASTStringDumper.h:
(WGSL::AST::dumpNode):
* Source/WebGPU/WGSL/AST/ASTTypeName.h:
* Source/WebGPU/WGSL/TypeCheck.cpp:
(WGSL::TypeChecker::TypeChecker):
(WGSL::TypeChecker::visit):
(WGSL::TypeChecker::infer):
(WGSL::TypeChecker::allocateType):
* Source/WebGPU/WGSL/Types.cpp: Added.
(WGSL::printInternal):
* Source/WebGPU/WGSL/Types.h: Added.
* Source/WebGPU/WebGPU.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/259805@main
